Return 0 if invalid seckey is given to ec_privkey_negate

jonasnick opened the pull request:
Fixes #488.

Instead of just fixing the documentation it's much better for misuse resistance to return 0 if the seckey overflows. I think most people who use this function and expect it to return 1 always will prefer their code to crash instead of getting a wrong result.

gmaxwell commented on src/tests.c:
```diff
@@ -4104,6 +4104,26 @@ void run_eckey_edge_case_test(void) {
+    CHECK(secp256k1_ec_privkey_negate(ctx, seckey) == 1);
+    CHECK(memcmp(seckey, seckey_tmp, 32) == 0);
+
+    /* Negating an overflowing seckey does not work */
```
Consistency of the actual behaviour would be better verified by a test that sets 16 bytes as 0xFF and then sets the remaining bytes randomly, in order to detect a change to a behaviour that  always sets the output to all 0xFF.

The documentation doesn't promise one behaviour vs the other, but preserving any entropy in the bad input is the safer behaviour (e.g. the data in it might get fed elsewhere into some rng hash or something) and the test should catch if it gets changed. 

A more extensive test would also(?) check the boundary conditions (smallest non-overflowing, smallest overflowing, maximum value). I don't think that's particularly important so long as scalar-set's overflow behaviour is checked precisely else where (I didn't look).

jonasnick replied:
I improved the test according to your suggestion. `scalar_set_b32` just uses `secp256k1_scalar_reduce` which I assume is well tested but I added a commit with boundary condition tests explicitly for `scalar_set_b32`.

real-or-random commented on include/secp256k1.h:
```diff
@@ -577,9 +577,11 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_create(
- *  In/Out: seckey:     pointer to the 32-byte private key to be negated (cannot be NULL)
+ *  In/Out: seckey:     pointer to the 32-byte private key to be negated. The private key
+ *                      interpreted as an integer (most significant byte first) must be less than
+ *                      the curve order. (cannot be NULL)
```
less than the curve order and not be 0?

(Actually we should just document what a valid private key is, e.g., here https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1.h#L552)

jonasnick replied:
It won't fail for 0 though. I think ideally we document what's a valid secret key is (as you said) in the doc of `seckey_verify` and then just refer to that instead of having to restate it over and over. I can rewrite this PR to do that.

jonasnick commented:
I added a commit to return 0 on an all 0 key because otherwise the result is an invalid seckey. Also moved definition of valid secret keys to verify_seckey as suggested by @real-or-random

real-or-random commented:
ACK 3f5512c81b17f1412d7992f8fe418a1ab629d955 I read the code and tested it, also with valgrind

It should be noted that this changes the observable behavior of ec_privkey_negate, which is also used in `CKey::Negate()` in Bitcoin Core, see https://github.com/bitcoin/bitcoin/blob/d0f81a96d9c158a9226dc946bdd61d48c4d42959/src/key.cpp#L168. But `CKey::Negate()` is only used in the tests, as far as I can tell.

jonasnick commented:
Superseded by #701
